src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs

```diff
@@ -192,7 +226,9 @@ internal void EndInvokeJS(long taskId, bool succeeded, ref Utf8JsonReader jsonRe
                 {
                     var resultType = TaskGenericsUtil.GetTaskCompletionSourceResultType(tcs);
 
+                    ByteArraysToDeserialize = byteArrays;
                     var result = JsonSerializer.Deserialize(ref jsonReader, resultType, JsonSerializerOptions);
+                    ByteArraysToDeserialize = null;
```

Comment:
Actually it's more than just thread-safety - it's also about nested invocations. I'm not sure whether it's possible in our current use cases, but it's imaginable that someone might have a JSON converter that itself performs JS interop.

If nested invocations is a scenario (not saying it is), then something like `StackObjectPool<T>` from the `Components` project might help with minimising the number of lists that get allocated, while retaining supported for nested usage.

Reply:
I had the same thoughts. I'm not concerned about thread safety since all these operations must happen inside the sync context. Nested invocations could be a thing, however I would suggest we detect that case and throw an exception as "unsupported" since it is possible but very unlikely.

Reply:
> I'm not concerned about thread safety since all these operations must happen inside the sync context

That may be true in our current uses but isn't part of the contract as far as the JS interop layer is concerned. JS interop is technically a separate thing and doesn't have a concept of a sync context.

We can define JSRuntime as being not thread-safe though, and yes it would be great to detect invalid usage and throw as this would simplify detecting any future issues.

Reply:
I believe https://github.com/dotnet/aspnetcore/pull/32259/commits/9995e90ba8d6192cfa80d2e23a569e1abcf319f6 & https://github.com/dotnet/aspnetcore/pull/32259/commits/8aeef7d17c919de7aba7b8f5a1671ab6e8919117 should resolve these concerns.

Regarding thread-safety/nested invocation: These added checks should be able to detect when we're trying to interrupt a serialization already in progress (and guard against it via semaphores which must be acquired prior to a (de)serialization request can be completed).

Please let me know if there are still additional concerns.